File: output/cpp/tests/31_doc_lamp_basic.cpp
#include <string>
#include <unordered_map>
#include <vector>
#include <any>
#include <memory>
#include <functional>


#include <iostream>
#include <string>
#include <cassert>

// Documentation Example: Basic Lamp with enter/exit events

class LampFrameEvent {
public:
    std::string _message;
    std::unordered_map<std::string, std::any> _parameters;

    LampFrameEvent(const std::string& message, std::unordered_map<std::string, std::any> params = {})
        : _message(message), _parameters(std::move(params)) {}
};

class LampFrameContext {
public:
    LampFrameEvent _event;
    std::any _return;
    std::unordered_map<std::string, std::any> _data;

    LampFrameContext(LampFrameEvent event, std::any default_return = {})
        : _event(std::move(event)), _return(std::move(default_return)) {}
};

class LampCompartment {
public:
    std::string state;
    std::unordered_map<std::string, std::any> state_args;
    std::unordered_map<std::string, std::any> state_vars;
    std::unordered_map<std::string, std::any> enter_args;
    std::unordered_map<std::string, std::any> exit_args;
    std::unique_ptr<LampFrameEvent> forward_event;
    std::unique_ptr<LampCompartment> parent_compartment;

    explicit LampCompartment(const std::string& state) : state(state) {}

    std::unique_ptr<LampCompartment> clone() const {
        auto c = std::make_unique<LampCompartment>(state);
        c->state_args = state_args;
        c->state_vars = state_vars;
        c->enter_args = enter_args;
        c->exit_args = exit_args;
        return c;
    }
};

class Lamp {
private:
    std::vector<std::unique_ptr<LampCompartment>> _state_stack;
    std::unique_ptr<LampCompartment> __compartment;
    std::unique_ptr<LampCompartment> __next_compartment;
    std::vector<LampFrameContext> _context_stack;

    void __kernel(LampFrameEvent& __e) {
        __router(__e);
        while (__next_compartment) {
            auto next_compartment = std::move(__next_compartment);
            LampFrameEvent exit_event("<$");
            __router(exit_event);
            __compartment = std::move(next_compartment);
            if (!__compartment->forward_event) {
                LampFrameEvent enter_event("$>");
                __router(enter_event);
            } else {
                auto forward_event = std::move(__compartment->forward_event);
                if (forward_event->_message == "$>") {
                    __router(*forward_event);
                } else {
                    LampFrameEvent enter_event("$>");
                    __router(enter_event);
                    __router(*forward_event);
                }
            }
        }
    }

    void __router(LampFrameEvent& __e) {
        const auto& state_name = __compartment->state;
        if (state_name == "Off") {
            _state_Off(__e);
        } else if (state_name == "On") {
            _state_On(__e);
        }
    }

    void __transition(std::unique_ptr<LampCompartment> next) {
        __next_compartment = std::move(next);
    }

    void _state_Off(LampFrameEvent& __e) {
        if (__e._message == "getColor") {
            _context_stack.back()._return = std::any(color);
            return;;
        } else if (__e._message == "isSwitchClosed") {
            _context_stack.back()._return = std::any(switch_closed);
            return;;
        } else if (__e._message == "setColor") {
            auto color = std::any_cast<std::string>(__e._parameters.at("color"));
            this->color = color;
        } else if (__e._message == "turnOn") {
            auto __new_compartment = std::make_unique<LampCompartment>("On");
            __new_compartment->parent_compartment = __compartment->clone();
            __transition(std::move(__new_compartment));
            return;
        }
    }

    void _state_On(LampFrameEvent& __e) {
        if (__e._message == "<$") {
            this->openSwitch();
        } else if (__e._message == "$>") {
            this->closeSwitch();
        } else if (__e._message == "getColor") {
            _context_stack.back()._return = std::any(color);
            return;;
        } else if (__e._message == "isSwitchClosed") {
            _context_stack.back()._return = std::any(switch_closed);
            return;;
        } else if (__e._message == "setColor") {
            auto color = std::any_cast<std::string>(__e._parameters.at("color"));
            this->color = color;
        } else if (__e._message == "turnOff") {
            auto __new_compartment = std::make_unique<LampCompartment>("Off");
            __new_compartment->parent_compartment = __compartment->clone();
            __transition(std::move(__new_compartment));
            return;
        }
    }

    void closeSwitch() {
                    switch_closed = true;
    }

    void openSwitch() {
                    switch_closed = false;
    }

public:
    std::string color = "white";
    bool switch_closed = false;

    Lamp() {
        __compartment = std::make_unique<LampCompartment>("Off");
        LampFrameEvent __frame_event("$>");
        LampFrameContext __ctx(std::move(__frame_event));
        _context_stack.push_back(std::move(__ctx));
        __kernel(_context_stack.back()._event);
        _context_stack.pop_back();
    }

    void turnOn() {
        LampFrameEvent __e("turnOn");
        LampFrameContext __ctx(std::move(__e));
        _context_stack.push_back(std::move(__ctx));
        __kernel(_context_stack.back()._event);
        _context_stack.pop_back();
    }

    void turnOff() {
        LampFrameEvent __e("turnOff");
        LampFrameContext __ctx(std::move(__e));
        _context_stack.push_back(std::move(__ctx));
        __kernel(_context_stack.back()._event);
        _context_stack.pop_back();
    }

    std::string getColor() {
        LampFrameEvent __e("getColor");
        LampFrameContext __ctx(std::move(__e), std::any(std::string()));
        _context_stack.push_back(std::move(__ctx));
        __kernel(_context_stack.back()._event);
        auto __result = std::any_cast<std::string>(std::move(_context_stack.back()._return));
        _context_stack.pop_back();
        return __result;
    }

    void setColor(std::string color) {
        std::unordered_map<std::string, std::any> __params;
        __params["color"] = color;
        LampFrameEvent __e("setColor", std::move(__params));
        LampFrameContext __ctx(std::move(__e));
        _context_stack.push_back(std::move(__ctx));
        __kernel(_context_stack.back()._event);
        _context_stack.pop_back();
    }

    bool isSwitchClosed() {
        LampFrameEvent __e("isSwitchClosed");
        LampFrameContext __ctx(std::move(__e), std::any(bool()));
        _context_stack.push_back(std::move(__ctx));
        __kernel(_context_stack.back()._event);
        auto __result = std::any_cast<bool>(std::move(_context_stack.back()._return));
        _context_stack.pop_back();
        return __result;
    }
};

int main() {
    printf("=== Test 31: Doc Lamp Basic ===\n");
    Lamp lamp;

    if (lamp.isSwitchClosed() != false) {
        printf("FAIL: Switch should be open initially\n");
        assert(false);
    }

    lamp.turnOn();
    if (lamp.isSwitchClosed() != true) {
        printf("FAIL: Switch should be closed after turnOn\n");
        assert(false);
    }

    if (lamp.getColor() != "white") {
        printf("FAIL: Expected 'white', got '%s'\n", lamp.getColor().c_str());
        assert(false);
    }

    lamp.setColor("blue");
    if (lamp.getColor() != "blue") {
        printf("FAIL: Expected 'blue', got '%s'\n", lamp.getColor().c_str());
        assert(false);
    }

    lamp.turnOff();
    if (lamp.isSwitchClosed() != false) {
        printf("FAIL: Switch should be open after turnOff\n");
        assert(false);
    }

    printf("PASS: Basic lamp works correctly\n");
    return 0;
}

File: output/go/tests/31_doc_lamp_basic.go

package main

import "fmt"
import "os"

type LampFrameEvent struct {
    _message    string
    _parameters map[string]any
}

type LampFrameContext struct {
    _event  LampFrameEvent
    _return any
    _data   map[string]any
}

type LampCompartment struct {
    state            string
    stateArgs        map[string]any
    stateVars        map[string]any
    enterArgs        map[string]any
    exitArgs         map[string]any
    forwardEvent     *LampFrameEvent
    parentCompartment *LampCompartment
}

func newLampCompartment(state string) *LampCompartment {
    return &LampCompartment{
        state:    state,
        stateArgs: make(map[string]any),
        stateVars: make(map[string]any),
        enterArgs: make(map[string]any),
        exitArgs:  make(map[string]any),
    }
}

func (c *LampCompartment) copy() *LampCompartment {
    nc := &LampCompartment{
        state: c.state,
        stateArgs: make(map[string]any),
        stateVars: make(map[string]any),
        enterArgs: make(map[string]any),
        exitArgs:  make(map[string]any),
        forwardEvent:     c.forwardEvent,
        parentCompartment: c.parentCompartment,
    }
    for k, v := range c.stateArgs { nc.stateArgs[k] = v }
    for k, v := range c.stateVars { nc.stateVars[k] = v }
    for k, v := range c.enterArgs { nc.enterArgs[k] = v }
    for k, v := range c.exitArgs { nc.exitArgs[k] = v }
    return nc
}

type Lamp struct {
    _state_stack []*LampCompartment
    __compartment *LampCompartment
    __next_compartment *LampCompartment
    _context_stack []LampFrameContext
    color string
    switch_closed bool
}

func NewLamp() *Lamp {
    s := &Lamp{}
    s._state_stack = make([]*LampCompartment, 0)
    s._context_stack = make([]LampFrameContext, 0)
    s.__compartment = newLampCompartment("Off")
    s.__next_compartment = nil
    __frame_event := LampFrameEvent{_message: "$>", _parameters: nil}
    __ctx := LampFrameContext{_event: __frame_event, _data: make(map[string]any)}
    s._context_stack = append(s._context_stack, __ctx)
    s.__kernel(&s._context_stack[len(s._context_stack)-1]._event)
    s._context_stack = s._context_stack[:len(s._context_stack)-1]
    return s
}

func (s *Lamp) __kernel(__e *LampFrameEvent) {
    s.__router(__e)
    for s.__next_compartment != nil {
        next_compartment := s.__next_compartment
        s.__next_compartment = nil
        exit_event := &LampFrameEvent{_message: "<$", _parameters: s.__compartment.exitArgs}
        s.__router(exit_event)
        s.__compartment = next_compartment
        if s.__compartment.forwardEvent == nil {
            enter_event := &LampFrameEvent{_message: "$>", _parameters: s.__compartment.enterArgs}
            s.__router(enter_event)
        } else {
            forward_event := s.__compartment.forwardEvent
            s.__compartment.forwardEvent = nil
            if forward_event._message == "$>" {
                s.__router(forward_event)
            } else {
                enter_event := &LampFrameEvent{_message: "$>", _parameters: s.__compartment.enterArgs}
                s.__router(enter_event)
                s.__router(forward_event)
            }
        }
    }
}

func (s *Lamp) __router(__e *LampFrameEvent) {
    switch s.__compartment.state {
    case "Off":
        s._state_Off(__e)
    case "On":
        s._state_On(__e)
    }
}

func (s *Lamp) __transition(next *LampCompartment) {
    s.__next_compartment = next
}

func (s *Lamp) TurnOn() {
    __e := LampFrameEvent{_message: "TurnOn"}
    __ctx := LampFrameContext{_event: __e, _data: make(map[string]any)}
    s._context_stack = append(s._context_stack, __ctx)
    s.__kernel(&s._context_stack[len(s._context_stack)-1]._event)
    s._context_stack = s._context_stack[:len(s._context_stack)-1]
}

func (s *Lamp) TurnOff() {
    __e := LampFrameEvent{_message: "TurnOff"}
    __ctx := LampFrameContext{_event: __e, _data: make(map[string]any)}
    s._context_stack = append(s._context_stack, __ctx)
    s.__kernel(&s._context_stack[len(s._context_stack)-1]._event)
    s._context_stack = s._context_stack[:len(s._context_stack)-1]
}

func (s *Lamp) GetColor() string {
    __e := LampFrameEvent{_message: "GetColor"}
    __ctx := LampFrameContext{_event: __e, _data: make(map[string]any)}
    s._context_stack = append(s._context_stack, __ctx)
    s.__kernel(&s._context_stack[len(s._context_stack)-1]._event)
    var __result string
    if __rv := s._context_stack[len(s._context_stack)-1]._return; __rv != nil { __result = __rv.(string) }
    s._context_stack = s._context_stack[:len(s._context_stack)-1]
    return __result
}

func (s *Lamp) SetColor(color string) {
    __params := map[string]any{
        "color": color,
    }
    __e := LampFrameEvent{_message: "SetColor", _parameters: __params}
    __ctx := LampFrameContext{_event: __e, _data: make(map[string]any)}
    s._context_stack = append(s._context_stack, __ctx)
    s.__kernel(&s._context_stack[len(s._context_stack)-1]._event)
    s._context_stack = s._context_stack[:len(s._context_stack)-1]
}

func (s *Lamp) IsSwitchClosed() bool {
    __e := LampFrameEvent{_message: "IsSwitchClosed"}
    __ctx := LampFrameContext{_event: __e, _data: make(map[string]any)}
    s._context_stack = append(s._context_stack, __ctx)
    s.__kernel(&s._context_stack[len(s._context_stack)-1]._event)
    var __result bool
    if __rv := s._context_stack[len(s._context_stack)-1]._return; __rv != nil { __result = __rv.(bool) }
    s._context_stack = s._context_stack[:len(s._context_stack)-1]
    return __result
}

func (s *Lamp) _state_On(__e *LampFrameEvent) {
    if __e._message == "<$" {
        s.__openSwitch()
    } else if __e._message == "$>" {
        s.__closeSwitch()
    } else if __e._message == "GetColor" {
        s._context_stack[len(s._context_stack)-1]._return = s.color
        return
    } else if __e._message == "IsSwitchClosed" {
        s._context_stack[len(s._context_stack)-1]._return = s.switch_closed
        return
    } else if __e._message == "SetColor" {
        color := __e._parameters["color"].(string)
        _ = color
        s.color = color
    } else if __e._message == "TurnOff" {
        __compartment := newLampCompartment("Off")
        __compartment.parentCompartment = s.__compartment.copy()
        s.__transition(__compartment)
        return
    }
}

func (s *Lamp) _state_Off(__e *LampFrameEvent) {
    if __e._message == "GetColor" {
        s._context_stack[len(s._context_stack)-1]._return = s.color
        return
    } else if __e._message == "IsSwitchClosed" {
        s._context_stack[len(s._context_stack)-1]._return = s.switch_closed
        return
    } else if __e._message == "SetColor" {
        color := __e._parameters["color"].(string)
        _ = color
        s.color = color
    } else if __e._message == "TurnOn" {
        __compartment := newLampCompartment("On")
        __compartment.parentCompartment = s.__compartment.copy()
        s.__transition(__compartment)
        return
    }
}

func (s *Lamp) __closeSwitch() {
                s.switch_closed = true
}

func (s *Lamp) __openSwitch() {
                s.switch_closed = false
}

func main() {
	fmt.Println("=== Test 31: Doc Lamp Basic ===")
	lamp := NewLamp()

	// Initially off - need to set color since Go zero-inits to ""
	// The Java version inits color to "white" via domain, but Go cannot do that
	// So we set it manually
	lamp.SetColor("white")

	if lamp.IsSwitchClosed() != false {
		fmt.Println("FAIL: Switch should be open initially")
		os.Exit(1)
	}

	// Turn on - should close switch
	lamp.TurnOn()
	if lamp.IsSwitchClosed() != true {
		fmt.Println("FAIL: Switch should be closed after TurnOn")
		os.Exit(1)
	}

	// Check color
	if lamp.GetColor() != "white" {
		fmt.Printf("FAIL: Expected 'white', got '%s'\n", lamp.GetColor())
		os.Exit(1)
	}

	// Set color
	lamp.SetColor("blue")
	if lamp.GetColor() != "blue" {
		fmt.Printf("FAIL: Expected 'blue', got '%s'\n", lamp.GetColor())
		os.Exit(1)
	}

	// Turn off - should open switch
	lamp.TurnOff()
	if lamp.IsSwitchClosed() != false {
		fmt.Println("FAIL: Switch should be open after TurnOff")
		os.Exit(1)
	}

	fmt.Println("PASS: Basic lamp works correctly")
}
